### PR TITLE
Add time namespace support to unshare

### PR DESCRIPTION
**Pull request by adrianreber**

### CODE REVIEW

**Comment by adrianreber** [comment]
@0x7f454c46 @avagin : As the authors of the kernel code just as an FYI. I hope I used the interfaces as you expected them to be used.

File: sys-utils/unshare.c [hunk 1]
```diff
@@ -284,6 +303,8 @@ static void __attribute__((__noreturn__)) usage(void)
 	fputs(_(" -w, --wd=<dir>            change working directory to <dir>\n"), out);
 	fputs(_(" -S, --setuid <uid>        set uid in entered namespace\n"), out);
 	fputs(_(" -G, --setgid <gid>        set gid in entered namespace\n"), out);
+	fputs(_(" --monotonic <offset>      set clock monotonic offset in time namespaces\n"), out);
+	fputs(_(" --boottime <offset>       set clock boottime offset in time namespaces\n"), out);
```
Comment (0x7f454c46) [inline sys-utils/unshare.c]: Nit: I'd probably clarify it by `--monotonic <offset seconds>`.

File: sys-utils/unshare.c [hunk 2]
```diff
@@ -213,6 +214,23 @@ static ino_t get_mnt_ino(pid_t pid)
 	return st.st_ino;
 }
 
+static void settime(time_t offset, clockid_t clk_id)
+{
+	char buf[4096];
```
Comment (karelzak) [inline sys-utils/unshare.c]: What about 
     char buf[ sizeof(stringify_value(ULONG_MAX)) * 3  ];
rathe rthan the magic number?

File: sys-utils/unshare.c [hunk 3]
```diff
@@ -443,6 +478,14 @@ int main(int argc, char *argv[])
 		case 'w':
 			newdir = optarg;
 			break;
+                case OPT_MONOTONIC:
+			monotonic = strtoul_or_err(optarg, _("failed to parse uid"));
```
Comment (karelzak) [inline sys-utils/unshare.c]: It seems like copy & past mistake. You probably want "failed to parse monotonic offset".

File: sys-utils/unshare.c [hunk 4]
```diff
@@ -443,6 +478,14 @@ int main(int argc, char *argv[])
...
+			force_monotonic = 1;
+			break;
+                case OPT_BOOTTIME:
+			boottime = strtoul_or_err(optarg, _("failed to parse uid"));
```
Comment (karelzak) [inline sys-utils/unshare.c]: "failed to parse bootime offset"

File: sys-utils/unshare.1 [hunk 5]
```diff
@@ -134,6 +141,10 @@ namespace is created by a bind mount.
 Unshare the cgroup namespace. If \fIfile\fP is specified then persistent namespace is created
 by bind mount.
 .TP
+.BR \-t , " \-\-time"[=\fIfile\fP]
+Unshare the time namespace. If \fIfile\fP is specified then a persistent
+namespace is created by a bind mount.
```
Comment (karelzak) [inline sys-utils/unshare.1]: maybe add a note about --monotonic and --boottime

File: sys-utils/unshare.1 [hunk 6]
```diff
@@ -208,6 +219,16 @@ Set the user ID which will be used in the entered namespace.
...
 .TP
+.BR "\-\-monotonic \fIoffset"
+Set the offset of
+.B CLOCK_MONOTONIC
+which will be used in the entered time namespace.
+.TP
+.BR "\-\-boottime \fIoffset"
+Set the offset of
+.B CLOCK_BOOTTIME
+which will be used in the entered time namespace.
```
Comment (karelzak) [inline sys-utils/unshare.1]: add a note that --time is required

**Comment by karelzak** [comment]
Looks good. Please, can you update the patch to fix the trivial cosmetic issues. Thanks!

File: sys-utils/unshare.c (hunk 1 above)
Comment (adrianreber) [reply sys-utils/unshare.c]: I included `seconds` in the help text as `<offset seconds>` did not fit in the help text formatting.

File: sys-utils/unshare.c (hunk 2 above)
Comment (adrianreber) [reply sys-utils/unshare.c]: done

File: sys-utils/unshare.c (hunk 4 above)
Comment (adrianreber) [reply sys-utils/unshare.c]: done

**Comment by adrianreber** [comment]
Updated patch to fix all review comments.

**Comment by karelzak** [comment]
Thanks! It seems last remaining place without "time" NS support is nsenter(1).

**Comment by adrianreber** [comment]
Thanks for mentioning `nsenter`. Let me check that.

**Comment by adrianreber** [comment]
@karelzak I see that nsenter already uses `-t` for target PID. Should we maybe switch to `-T` for `unshare` to have the same command-line parameters in nsenter and unshare?

**Comment by karelzak** [comment]
Hmm... good point, it would be nice to have the same in all places. Send patch :-)

**Pull request merged**